syft/execution/protocol.py

```diff
@@ -1,479 +1,560 @@
+        # List of available translations
+        # self.translations = []
 
-        self._resolve_workers(workers)
+    # @property
+    # def state(self):
+    #     return self.role.state
 
-        # update plans list with *pointers* to the plans
-        self.plans = [(worker, plan.send(worker)) for (worker, plan) in self.plans]
+    # def parameters(self):
+    #     """
+    #     This is defined to match the torch api of nn.Module where .parameters() return the model tensors / parameters
+    #     """
+    #     if self.state is not None:
+    #         return self.state.tensors()
+    #     else:
+    #         return []
 
-        return self
+    def get_role_for_owner(self, owner):
+        if owner.id not in self.roles:
+            self.roles[owner.id] = Role()
+        return self.roles[owner.id]
+
+    def build(self, *args):
+        """Builds the protocol.
 
-    def __call__(self, *args, **kwargs):
-        return self.run(*args, **kwargs)
+        First, run the function to be converted in a protocol in a context which
+        activates the tracing and record the actions in trace.logs
+
+        Second, store the result ids temporarily to helper ordering the output
+        placeholders at return time
+
+        Third, loop through the trace logs and replace the tensors found in the
+        actions logged by PlaceHolders. Record those actions in
+        protocol.actions
+
+        Args:
+            args: Input arguments to run the protocol
+        """
+        # Reset previous build
+        self.roles = {}
+
+        # Enable tracing
+        self.toggle_tracing(True)
+        self.is_building = True
+
+        # Run once to build the protocol
+        ph_args = tuple()
+        for arg in args:
+            arg_role = self.get_role_for_owner(arg.owner)
+
+            ph_arg = PlaceHolder.create_from(arg, owner=arg.owner, role=arg_role, tracing=True)
+            # Register inputs in role
+            arg_role.register_input(ph_arg)
+
+            ph_args += (ph_arg,)
+
+        # Add state to args if needed
+        # if self.include_state:
+        #     ph_args += (self.state,)
+
+        # with trace(framework_packages["torch"], self.role, self.owner) as wrapped_torch:
+        # Look for framework kwargs
+        # framework_kwargs = {}
+        # forward_args = inspect.getargspec(self.forward).args
+        # if "torch" in forward_args:
+        #     framework_kwargs["torch"] = wrapped_torch
+
+        # results = self.forward(*args, **framework_kwargs)
+        results = self.forward(*ph_args)
+
+        # Disable tracing
+        self.toggle_tracing(False)
+        self.is_building = False
+
+        # Register outputs in roles
+        for result in results:
+            if isinstance(result, PlaceHolder):
+                result_role = self.get_role_for_owner(result.owner)
+                result_role.register_output(result)
+
+        self.is_built = True
+
+        # Build registered translations
```

Comment:
Same with translations: we'll want them eventually, but I think it's fine to completely remove them for now and add them back later.

Reply:
I'll remove them